typing/path.ml

```diff
@@ -55,71 +85,59 @@ let exists_free ids p =
 
 let rec scope = function
     Pident id -> Ident.scope id
-  | Pdot(p, _s) -> scope p
+  | Pdot(p, _) | Pextra_ty (p, _) -> scope p
   | Papply(p1, p2) -> Int.max (scope p1) (scope p2)
 
 let kfalse _ = false
 
 let rec name ?(paren=kfalse) = function
     Pident id -> Ident.name id
-  | Pdot(p, s) ->
+  | Pdot(p, s) | Pextra_ty (p, Pcstr_ty s) ->
       name ~paren p ^ if paren s then ".( " ^ s ^ " )" else "." ^ s
   | Papply(p1, p2) -> name ~paren p1 ^ "(" ^ name ~paren p2 ^ ")"
+  | Pextra_ty (p, Pext_ty) -> name ~paren p
+  | Pextra_ty (p, Pcls_ty) -> "#" ^ name ~paren p
 
 let rec print ppf = function
   | Pident id -> Ident.print_with_scope ppf id
-  | Pdot(p, s) -> Format.fprintf ppf "%a.%s" print p s
+  | Pdot(p, s) | Pextra_ty (p, Pcstr_ty s) ->
+      Format.fprintf ppf "%a.%s" print p s
   | Papply(p1, p2) -> Format.fprintf ppf "%a(%a)" print p1 print p2
+  | Pextra_ty (p, Pext_ty) -> print ppf p
+  | Pextra_ty (p, Pcls_ty) -> Format.fprintf ppf "#%a" print p
 
 let rec head = function
     Pident id -> id
-  | Pdot(p, _s) -> head p
+  | Pdot(p, _) | Pextra_ty (p, _) -> head p
   | Papply _ -> assert false
 
 let flatten =
   let rec flatten acc = function
     | Pident id -> `Ok (id, acc)
-    | Pdot (p, s) -> flatten (s :: acc) p
+    | Pdot (p, s) | Pextra_ty (p, Pcstr_ty s) -> flatten (s :: acc) p
     | Papply _ -> `Contains_apply
+    | Pextra_ty (p, Pext_ty) -> flatten acc p
+    | Pextra_ty (p, Pcls_ty) -> flatten ("#" :: acc) p
   in
   fun t -> flatten [] t
 
 let heads p =
   let rec heads p acc = match p with
     | Pident id -> id :: acc
-    | Pdot (p, _s) -> heads p acc
+    | Pdot (p, _) | Pextra_ty (p, _) -> heads p acc
     | Papply(p1, p2) ->
         heads p1 (heads p2 acc)
   in heads p []
 
 let rec last = function
   | Pident id -> Ident.name id
-  | Pdot(_, s) -> s
-  | Papply(_, p) -> last p
-
-let is_uident s =
-  assert (s <> "");
-  match s.[0] with
-  | 'A'..'Z' -> true
-  | _ -> false
-
-type typath =
-  | Regular of t
-  | Ext of t * string
-  | LocalExt of Ident.t
-  | Cstr of t * string
-
-let constructor_typath = function
-  | Pident id when is_uident (Ident.name id) -> LocalExt id
-  | Pdot(ty_path, s) when is_uident s ->
-      if is_uident (last ty_path) then Ext (ty_path, s)
-      else Cstr (ty_path, s)
-  | p -> Regular p
+  | Pdot(_, s) | Pextra_ty (_, Pcstr_ty s) -> s
+  | Papply(_, p) | Pextra_ty (p, (Pext_ty | Pcls_ty)) -> last p
```

Comment:
Shouldn't it be `"#" ^ last p` for `Pcls_ty` ?
At this point `last` is never called on such a path, so we cannot see the difference.

Reply:
Right, I’ll fix it. 
I would be better to add some test cases for it. I’ll try it.

Reply:
For now, I fixed the code part. I'll keep investigating to add appropriate test.